Lib/exception_group.py

```diff
@@ -0,0 +1,249 @@
+
+        excs: sequence of exceptions
+        tb [optional]: the __traceback__ of this exception group.
+        Typically set when this ExceptionGroup is derived from another.
+        """
+        self.excs = excs
+        self.message = message
+        super().__init__(self.message)
+        # self.__traceback__ is updated as usual, but self.__traceback_group__
+        # is set when the exception group is created.
+        # __traceback_group__ and __traceback__ combine to give the full path.
+        self.__traceback__ = tb
+        self.__traceback_group__ = TracebackGroup(self.excs)
+
+    def project(self, condition, with_complement=False):
+        """ Split an ExceptionGroup based on an exception predicate
+
+        returns a new ExceptionGroup, match, of the exceptions of self
+        for which condition returns True. If with_complement is True,
+        returns another ExceptionGroup for the exception for which
+        condition returns False.
+        match and rest have the same nested structure as self, but empty
+        sub-exceptions are not included. They have the same message,
+        __traceback__, __cause__ and __context__ fields as self.
+
+        condition: BaseException --> Boolean
+        with_complement: Bool  If True, construct also an EG of the non-matches
+        """
+        match = []
+        rest = [] if with_complement else None
+        for e in self.excs:
+            if isinstance(e, ExceptionGroup): # recurse
+                e_match, e_rest = e.project(
+                    condition, with_complement=with_complement)
+                if not e_match.is_empty():
+                    match.append(e_match)
+                if with_complement and not e_rest.is_empty():
+                    rest.append(e_rest)
+            else:
+                if condition(e):
+                    match.append(e)
+                elif with_complement:
+                    rest.append(e)
+
+        match_exc = ExceptionGroup(match, tb=self.__traceback__)
+        def copy_metadata(src, target):
+            target.message = src.message
+            target.__context__ = src.__context__
+            target.__cause__ = src.__cause__
+        copy_metadata(self, match_exc)
+        if with_complement:
+            rest_exc = ExceptionGroup(rest, tb=self.__traceback__)
+            copy_metadata(self, rest_exc)
+        else:
+            rest_exc = None
+        return match_exc, rest_exc
+
+    def split(self, type):
+        """ Split an ExceptionGroup to extract exceptions of type E
+
+        type: An exception type
+        """
+        return self.project(
+            lambda e: isinstance(e, type),
+            with_complement=True)
+
+    def subgroup(self, keep):
+        """ Split an ExceptionGroup to extract only exceptions in keep
+
+        keep: List[BaseException]
+        """
+        match, _ = self.project(lambda e: e in keep)
+        return match
+
+    def extract_traceback(self, exc):
```

Comment:
Not sure we need this as part of the API. It's used in the tests and can move there.